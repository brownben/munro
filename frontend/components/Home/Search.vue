<script setup lang="ts">
import { MagnifyingGlassIcon } from '@heroicons/vue/24/outline'
</script>
<template>
  <section class="bg-main-600 dark:bg-main-800 text-white">
    <div
      class="mx-auto flex max-w-(--breakpoint-lg) flex-col gap-2 px-6 py-10 text-center sm:py-16 lg:px-8"
    >
      <h2 class="text-main-100 text-4xl font-medium md:text-5xl">
        Find <strong class="font-black text-white">Everything</strong> You Need
      </h2>
      <p class="text-lg md:text-xl">
        Instantly search all leagues, events and competitors
      </p>
      <form action="/search" method="get" class="mt-6 flex justify-center">
        <label for="search-box" class="sr-only">Search Query:</label>
        <div class="relative w-full sm:w-1/2">
          <input
            id="search-box"
            name="query"
            type="search"
            class="ring-offset-main-600 w-full rounded-sm border border-white bg-transparent p-2 pr-16 ring-white ring-offset-2 outline-hidden focus:ring-3"
          />
          <button
            class="hover:text-main-600 focus:text-main-600 absolute top-0 right-0 h-full rounded-sm px-4 outline-hidden transition hover:bg-white focus:bg-white"
          >
            <span class="sr-only">Search</span>
            <MagnifyingGlassIcon class="h-6 w-6" />
          </button>
        </div>
      </form>
    </div>
  </section>
</template>
