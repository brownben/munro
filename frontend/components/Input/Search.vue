<script setup lang="ts">
import { MagnifyingGlassIcon } from '@heroicons/vue/24/outline/index.js'
</script>
<template>
  <form method="get" action="/search" class="flex justify-center sm:mb-4">
    <label for="search-box" class="sr-only">Search Query:</label>
    <div class="group relative w-full">
      <input
        id="search-box"
        name="query"
        type="search"
        class="w-full rounded border border-gray-300 bg-transparent p-2 pr-12 text-sm outline-none ring-main-600 ring-offset-2 focus:border-main-600 focus:ring"
        :value="$route.query.query"
      />
      <button
        class="absolute top-0 right-0 h-full rounded px-3 text-gray-400 outline-none transition hover:bg-main-600 hover:text-white focus:bg-main-600 group-focus-within:text-main-600 group-focus-within:hover:text-white group-focus-within:focus:text-white"
      >
        <span class="sr-only">Search</span>
        <MagnifyingGlassIcon class="h-5 w-5" />
      </button>
    </div>
  </form>
</template>
