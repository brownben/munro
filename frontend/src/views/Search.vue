<template>
  <Layout has-mobile-sub-title gray>
    <template #title>
      <Meta
        :head="{ meta: { name: 'robots', content: 'all' } }"
        :title="`Munro - &quot;${$route.params.query || ''}&quot;`"
        description="Munro - League Results.
    Sorted. Sports League Results Calculated Quick and Easily, with Results
    Sorting and Filtering Options"
        url="https://munro-leagues.herokuapp.com/search"
      />
      <div
        class="flex flex-col items-start justify-between sm:items-center sm:flex-row"
      >
        <h1 class="mb-4 text-3xl font-bold leading-tight sm:mb-0 font-heading">
          Search
        </h1>
        <SearchBox />
      </div>
    </template>

    <div
      v-if="leagues.length > 0"
      class="flex flex-row flex-wrap justify-between w-full col-span-2 -mb-4"
    >
      <h2
        class="inline-block py-2 leading-tight tracking-wide uppercase select-none font-heading text-main-600"
      >
        Leagues
      </h2>
      <router-link
        to="/leagues"
        class="inline-block px-4 pt-2 text-xs leading-6 tracking-wider text-right uppercase transition duration-300 select-none text-main-600 font-heading hover:bg-main-100 focus:bg-main-100 rounded-shape"
      >
        View All Leagues â†’
      </router-link>
    </div>

    <LeagueOverviewCard
      v-for="league of leagues"
      :key="league.name"
      :title="league.name"
      :description="league.description"
    />

    <h2
      v-if="events.length > 0"
      class="col-span-2 -mb-4 uppercase select-none font-heading text-main-700"
    >
      Events
    </h2>

    <EventOverviewCard
      v-for="event of events"
      :key="event.id"
      :event="event"
      :league="{ dynamicEventResults: true }"
      :show-league-name="true"
      :show-full-details="false"
    />

    <h2
      v-if="competitors.length > 0"
      class="col-span-2 -mb-4 uppercase select-none font-heading text-main-700"
    >
      Competitors
    </h2>

    <CompetitorOverviewCard
      v-for="competitor of competitors"
      :key="competitor.id"
      :competitor="competitor"
    />
    <NoResults
      v-if="
        leagues.length === 0 &&
        events.length === 0 &&
        competitors.length === 0 &&
        !loading &&
        $route.params.query
      "
      :text="`Sorry I Couldn't Find &quot;${$route.params.query || ''}&quot;`"
      secondary-text="Try Tweaking Your Search Query To Find a Match"
      class="col-span-2"
    />
    <p v-else>&nbsp;</p>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, watch, onMounted, computed, defineAsyncComponent } from 'vue'
import { useRouter, useRoute } from 'vue-router'

import Layout from '../components/Layout.vue'
import SearchBox from '../components/inputs/SearchBox.vue'
import EventOverviewCard from '../components/cards/EventOverviewCard.vue'
import LeagueOverviewCard from '../components/cards/LeagueOverviewCard.vue'
import CompetitorOverviewCard from '../components/cards/CompetitorOverviewCard.vue'

const NoResults = defineAsyncComponent(
  () => import('../components/cards/NoResultsCard.vue')
)

import { toSingleString } from '../scripts/typeHelpers'

import { getQuery } from '../api/search'

const router = useRouter()
const route = useRoute()

const loading = ref(true)
const leagues = ref<League[]>([])
const events = ref<Event[]>([])
const competitors = ref<Competitor[]>([])

const getDetails = async () => {
  loading.value = true
  const routeParamsQuery = toSingleString(route.params.query || '')
  const queryResult = await getQuery(routeParamsQuery)
  if (queryResult) {
    leagues.value = queryResult.leagues
    events.value = queryResult.events
    competitors.value = queryResult.competitors
  }
  loading.value = false
}

watch(route, getDetails, { immediate: true })
</script>
