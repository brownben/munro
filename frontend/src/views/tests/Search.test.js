import { mount, shallowMount } from '@vue/test-utils'
import Search from '@/views/Search.vue'

import axios from 'axios'

const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

jest.mock('axios')

beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockRejectedValue()
})

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Search, {
    stubs: ['router-link'],
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = shallowMount(Search, {
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link'],
  })
  wrapper.setData({ loading: true, leagues: [], events: [], competitors: [] })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - No Results', () => {
  const wrapper = mount(Search, {
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ loading: false, leagues: [], events: [], competitors: [] })
  expect(wrapper.element).toMatchSnapshot()
})

test('Get Details - Request Called Correctly', async () => {
  const wrapper = mount(Search, {
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getDetails()

  expect(axios.get).toHaveBeenLastCalledWith('/api/search?query=a')
})

test('Get Details - Processes Data', async () => {
  const wrapper = shallowMount(Search, {
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({
    data: {
      leagues: [{ l: 'l' }],
      events: [{ e: 'e' }],
      competitors: [{ c: 'c' }],
    },
  })
  await wrapper.vm.getDetails()
  expect(wrapper.vm.leagues).toEqual([{ l: 'l' }])
  expect(wrapper.vm.events).toEqual([{ e: 'e' }])
  expect(wrapper.vm.competitors).toEqual([{ c: 'c' }])
  expect(wrapper.vm.loading).toBe(false)

  expect(axios.get).toHaveBeenLastCalledWith('/api/search?query=a')
})

test('Get Details - Shows Message on Error', async () => {
  const wrapper = mount(Search, {
    mocks: {
      $route: { params: { query: 'a' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getDetails()
  expect(wrapper.vm.leagues).toEqual([])

  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Data'
  )
})
