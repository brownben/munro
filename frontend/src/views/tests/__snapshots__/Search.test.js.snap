// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - No Results 1`] = `
<div
  class="flex flex-col flex-grow h-full"
>
  <header
    class="text-gray-900 bg-white"
  >
    <div
      class="max-w-screen-xl px-6 pb-6 mx-auto lg:px-8 pt-3 md:pt-6"
    >
      <vue-headful-stub
        description="Munro - League Results.
  Sorted. Sports League Results Calculated Quick and Easily, with Results
  Sorting and Filtering Options"
        head="[object Object]"
        title="Munro - \\"a\\""
        url="https://munro-leagues.herokuapp.com/search"
      />
       
      <div
        class="flex flex-col items-start justify-between sm:items-center sm:flex-row"
      >
        <h1
          class="mb-4 text-3xl font-bold md:mb-0"
        >
          Search
        </h1>
         
        <form
          class="relative w-full sm:w-1/2 lg:w-1/3"
        >
          <label
            class="sr-only"
            for="searchBox"
          >
            Search Query:
          </label>
           
          <input
            class="w-full py-2 pl-10 pr-24 font-sans leading-snug text-gray-500 transition duration-300 ease-in-out border border-gray-100 outline-none appearance-none bg-gray-50 rounded-shape focus:border-main-300 focus:text-gray-900 focus:bg-white focus:shadow-outline"
            id="searchBox"
            spellcheck="false"
            type="text"
          />
           
          <button
            class="absolute top-0 right-0 px-4 py-2 transition duration-300 outline-none font-heading rounded-shape text-gray-600 bg-gray-100 hover:bg-main-200 focus:bg-main-200 hover:text-main-700 focus:text-main-700"
          >
            
    Search
  
          </button>
           
          <svg
            class="absolute top-0 left-0 w-6 h-6 py-px my-2 ml-2 text-gray-400"
            fill="currentColor"
            viewBox="0 0 20 20"
          >
            <path
              clip-rule="evenodd"
              d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
              fill-rule="evenodd"
            />
          </svg>
        </form>
      </div>
    </div>
  </header>
   
  <main
    class="flex flex-col flex-grow"
  >
    <!---->
     
    <!---->
     
    <div
      class="flex-grow bg-gray-50 border-t border-main-100 pt-8"
    >
      <div
        class="max-w-screen-xl pb-8 mx-auto sm:pb-10 xl:pb-12 lg:px-8 sm:px-6"
      >
        <div
          class="grid grid-cols-2 gap-6 px-6 sm:px-0 sm:gap-8"
        >
           
          <!---->
            
          <!---->
            
          <!---->
            
          <!---->
        </div>
      </div>
    </div>
  </main>
   
  <!---->
</div>
`;

exports[`Renders Correctly 1`] = `
<layout-stub
  gray="true"
  hasmobilesubtitle="true"
  title=""
>
   
  <!---->
    
  <!---->
    
  <!---->
    
  <p>
    Â 
  </p>
</layout-stub>
`;
